### PR TITLE
Use CSS grid in `Section`

### LINKED ISSUE #6987: Remove padding-left and -right on the cards at edge so they sit on the grid
## Issue

DCR fronts do not sit well in the [Guardian grid](https://theguardian.design/2a1e5182b/p/41be19-grids) in smaller break points.

| frontend | DCR |
|----------|-----|
| image | image |
| image | image |

There was an [attempt](https://github.com/guardian/dotcom-rendering/pull/6993) to fix it without refactoring but we had to [revert](https://github.com/guardian/dotcom-rendering/pull/7044).

Inspired by the [work done in AR that implemented Guardian grid using CSS grid](https://github.com/guardian/dotcom-rendering/pull/4874) we aim to do the same for article layouts and fronts layout.

### Benefits
* It will be a lot easier to fix the current bug and manipulate the layouts in the future
* This will reduce layout shift, hence also contribute to the work the Open Journalism team does.  Here's an [interesting video](https://www.youtube.com/watch?v=7EKEav7Io5Y) that explains how and why.

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

This is a major refactor of `Section` to use [CSS grid](https://developer.mozilla.org/en-US/docs/Web/CSS/grid), via the existing AR helper, [`grid.ts`](https://github.com/guardian/dotcom-rendering/blob/8e998546b0290a3035441b56fe89bd334bd1fb84/apps-rendering/src/grid/grid.ts#L15). Instead of relying on the `flex` layout algorithm, we rely on `grid` to ensure strict adherence to the [Guardian’s vertical grid](https://theguardian.design/2a1e5182b/p/41be19-grids).

Fixes the following issues:

- When a containter title is too long, [it does not expand the left column](https://github.com/guardian/dotcom-rendering/pull/7128#issuecomment-1422375286)
- Below tablet, the date line is not visible below the container title
- Treats get hidden along with the content when a container is hidden.

## Why?

This brings the design intent and CSS much closer together, as we now have exactly 4, 12, 14 and 16 columns at the different breakpoints and content does not have to introduce complex padding logic to account for the gutters.

The DOM can also be a lot shallower as all the grid elements are siblings.

Closes #6987

## Examples

<details>
<summary>from `mobile` (320) to `phablet` (660)</summary>
<pre>
 1 2 3 4
┌─────┬─┐
│Title│X│
├─────┴─┤
│▒▒▒▒▒▒▒│
│▒▒▒▒▒▒▒│
└───────┘
</pre>
</details>

<details>
<summary>from `tablet` (740) to `desktop` (980)</summary>
<pre>
 1 2 3 4 5 6 7 8 9 a b c (12)
┌─────────────────────┬─┐
│Title                │X│
│Date                 │ │
├─────────────────────┴─┤
│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
└───────────────────────┘
</pre>
</details>

<details>
<summary>on `leftCol` (1140) if component is toggleable</summary>
<pre>
 1 2 3 4 5 6 7 8 9 a b c d e (14)
┌───┬───────────────────┬─┐
│Tit│                   │X│
│Dat├───────────────────┴─┤
├───┤▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│   │▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│Tre│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│ats│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
└───┴─────────────────────┘
</pre>
</details>

<details>
<summary>on `leftCol` (1140) if component is not toggleable</summary>
<pre>
 1 2 3 4 5 6 7 8 9 a b c d e (14)
┌───┬──────────────────────┐
│Tit│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│Dat│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
├───┤▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│   │▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│Tre│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│ats│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
└───┴──────────────────────┘
</pre>
</details>

<details>
<summary>on `wide` (1300)</summary>
<pre>
 1 2 3 4 5 6 7 8 9 a b c d e f g (16)
┌─────┬───────────────────────┬─┐
│Title│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│X│
│Date │▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒└─┤
├─────┤▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  │
│     │▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  │
│     │▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  │
│Treat│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  │
└─────┴─────────────────────────┘
</pre>
</details>

<details>
<summary>on `wide` (1300) with `stretchRight`</summary>
<pre>
 1 2 3 4 5 6 7 8 9 a b c d e f g (16)
┌─────┬─────────────────────────┐
│Title│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│Date │▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
├─────┤▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│     │▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│     │▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
│Treat│▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒│
└─────┴─────────────────────────┘
</pre>
</details>

## Screenshots

| Before      | After      |
|-------------|------------|
| ![before][] | ![after][] |
| ![date mobile before][] | ![date mobile after][] |
| ![discussion before][] | ![discussion after][] |

[before]: https://user-images.githubusercontent.com/76776/216674739-1b9e3656-ea49-45d2-9210-a9c41e4abb94.png
[after]: https://user-images.githubusercontent.com/76776/216674656-3a2ad2e7-dcb4-43b1-9990-4da1488fc7cb.png

[date mobile before]: https://user-images.githubusercontent.com/76776/217257416-95e0a378-ff26-4fba-927d-1f9ef0b0e4f4.png
[date mobile after]: https://user-images.githubusercontent.com/76776/217257400-c1f08253-4df5-49d7-8095-04692e28511f.png

[discussion before]: https://user-images.githubusercontent.com/76776/217257644-10a24809-d610-477b-b03b-b663180ad87f.png
[discussion after]: https://user-images.githubusercontent.com/76776/217257624-635b51e8-5c03-446c-9eee-df60a0a30deb.png

### CODE REVIEW

**Comment by HarryFischer** [comment]
I think the MPU advert on an article should have a column space between the text column and ad?

**Comment by ioannakok** [comment]
On fronts all breakpoints look great apart from desktop which is a bit off:

image

File: dotcom-rendering/src/lib/grid.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+import { from } from '@guardian/source-foundations';
+
+// ----- Columns & Lines ----- //
+
+/**
+ * Named CSS grid lines, based on the three columns commonly used for Guardian
+ * layouts.
+ */
+type Line =
```
Comment (ioannakok) [inline dotcom-rendering/src/lib/grid.ts:11]: This API is already used in AR for Immersives and Galleries: https://github.com/guardian/dotcom-rendering/pull/4874

File: dotcom-rendering/src/lib/grid.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ * @param start The grid line to start from, either a `Line` name or a number.
+ * @param span The number of columns to span.
+ * @returns {string} CSS to place the element on the grid.
+ *
+ * @example <caption>The element will span 3 columns from the line.</caption>
+ * const styles = css`
+ *   ${grid.span('centre-column-start', 3)}
+ * `;
+ */
+const span = (start: Line | number, span: number): string => `
```
Comment (bryophyta) [inline dotcom-rendering/src/lib/grid.ts:97]: I appreciate the type safety that we get for named lines with this API, but I'm in two minds about the pattern. I think part of my discomfort comes from the fact that `span` is already term in the CSS spec, but we're using it here as an alias for `grid-column`. We could rename the function, but part of me thinks that we should encourage devs to use the CSS spec directly when it comes to placing elements on the grid.

There's clear value to a reusable abstraction for _creating_ the grid containers, but for placing elements on it once it's been created I'm not sure whether type safety is worth the extra abstraction for `span` and `between` etc.?

**Comment by ioannakok** [comment]
Sorry deleted previous comment for desktop being a bit off. Had previous version of the code 🤦‍♀️

**Comment by ioannakok** [comment]
Some suspicious stories:

https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7c
https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7f
https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc28
https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc29

**Comment by mxdvl** [comment]
> Some suspicious stories:
> 
> [chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7c](https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7c) [chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7f](https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7f) [chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc28](https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc28) [chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc29](https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc29)


These were all originally wrong! The left column was wider than expected, because it was allowed to grow as part of the flex layout, as the title was too long to fit in the title. We would not want a long section title to break our grid, so the baseline is actually not as per our designs.

**Comment by ioannakok** [comment]
> > Some suspicious stories:
> > [chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7c](https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7c) [chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7f](https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850db7f) [chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc28](https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc28) [chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc29](https://www.chromatic.com/test?appId=63e251470cfbe61776b0ef19&id=63e27ee5d6b8c3995850dc29)
> 
> These were all originally wrong! The left column was wider than expected, because it was allowed to grow as part of the flex layout, as the title was too long to fit in the title. We would not want a long section title to break our grid, so the baseline is actually not as per our designs.

OK I see, makes sense. What do you think about breaking the text in multiple lines in these stories, e.g. `EventAlt Palette`?

File: dotcom-rendering/src/lib/grid.ts (hunk 2 above)
Comment (mxdvl) [reply dotcom-rendering/src/lib/grid.ts:97]: I think that if these abstractions become helpful when you use them a lot. Maybe for this Section we don’t need them, and I’ll have a look at a pure CSS implementation.

**Comment by mxdvl** [comment]
> What do you think about breaking the text in multiple lines in these stories, e.g. `EventAlt Palette`?

I think that's a fix forward, this PR is already a big refactor

**Review by ioannakok — APPROVED** [review]
Amazing! Looks great to me!

File: dotcom-rendering/src/web/components/Section.tsx [hunk 3]
```diff
@@ -107,78 +103,238 @@ type Props = {
...
+	display: grid;
+	grid-template-columns:
+		[viewport-start] 0px
+		[content-start title-start]
+		repeat(3, minmax(0, 1fr))
+		[title-end hide-start]
+		minmax(0, 1fr)
+		[content-end hide-end]
+		0px [viewport-end];
+	column-gap: 12px;
```
Comment (HarryFischer) [inline dotcom-rendering/src/web/components/Section.tsx:115]: Can we change this to 10px please

Comment (mxdvl) [reply dotcom-rendering/src/web/components/Section.tsx]: Done!

File: dotcom-rendering/src/lib/grid.ts (hunk 1 above)
Comment (mxdvl) [reply dotcom-rendering/src/lib/grid.ts]: We went with a CSS only version for the time being.

File: dotcom-rendering/src/lib/grid.ts (hunk 2 above)
Comment (mxdvl) [reply dotcom-rendering/src/lib/grid.ts]: Merged #7146 which implements a pure CSS implementation.

**Pull request merged**